make_prg/utils/msa_aligner.py

```diff
@@ -0,0 +1,137 @@
+import uuid
+import shutil
+from pathlib import Path
+from typing import Set
+import time
+import subprocess
+from loguru import logger
+import os
+from Bio.AlignIO import MultipleSeqAlignment
+from Bio import SeqIO
+from make_prg.utils.io_utils import load_alignment_file
+
+
+class TempDirAlreadyExistsError(Exception):
+    pass
+
+
+class NotAValidExecutableError(Exception):
+    pass
+
+
+class ExecutionError(Exception):
+    pass
+
+
+class MSAAligner(ABC):
+    def _set_executable(self, executable: str):
+        is_valid_executable = shutil.which(executable, mode=os.X_OK) is not None
+        if not is_valid_executable:
+            raise NotAValidExecutableError(f"Given MSA executable {executable} does not work or is invalid")
+        self._executable: str = executable
+
+    def _set_tmpdir(self, tmpdir: Path):
+        tmpdir.mkdir(parents=True, exist_ok=True)
+        self._tmpdir: Path = tmpdir
+
+    def __init__(self, executable: str, tmpdir: Path = Path("..")):
+        self._set_executable(executable)
+        self._set_tmpdir(tmpdir)
+
+    @abstractmethod
+    def get_updated_alignment(self, current_alignment: MultipleSeqAlignment, new_sequences: Set[str]) -> MultipleSeqAlignment:
+        pass
+
+    @classmethod
+    def get_aligner_name(cls) -> str:
+        pass
+
+    def _run_aligner(self, args: str, env=None):
+        start = time.time()
+        process = subprocess.Popen(
+            args,
+            stderr=subprocess.PIPE,
+            encoding="utf-8",
+            shell=True,
+            env=env,
+        )
+        exit_code = process.wait()
+
+        if exit_code != 0:
+            raise ExecutionError(
+                f"Failed to execute {self.__class__.get_aligner_name()} for arguments {args} due to the following "
+                f"error:\n{process.stderr.read()}"
+            )
+        stop = time.time()
+        runtime = stop - start
+        logger.debug(f"{self.__class__.get_aligner_name()} runtime for arguments {args} in seconds: {runtime:.3f}")
+
+    @staticmethod
+    def _create_new_sequences_file(directory: Path, new_sequences: Set[str]) -> Path:
+        # this is just done so that we have a deterministic order of new_sequences and tests run correctly
+        new_sequences = sorted(list(new_sequences))
+        new_sequences_filepath = directory / "new_sequences.fa"
+        with open(new_sequences_filepath, "w") as new_sequences_handler:
+            for index_new_seq, new_seq in enumerate(new_sequences):
+                print(
+                    f">Denovo_path_{index_new_seq}",
+                    file=new_sequences_handler,
+                )
+                print(new_seq, file=new_sequences_handler)
+        return new_sequences_filepath
+
+
+class MAFFT(MSAAligner):
+    @classmethod
+    def get_aligner_name(cls) -> str:
+        return "MAFFT"
+
+    def _prepare_run_tmpdir(self) -> Path:
+        run_tmpdir = self._tmpdir / f"temp.MSA.{uuid.uuid4()}"
```

Comment:
Again, why are we creating temporary directories? There's much less chance of things going wrong if we use `tempfile`?

Reply:
Temporary dirs are needed for three reasons. 1. When we multiprocess, each process create 5 files per locus. To avoid filesystem bloat, we create one dir per locus, and put the 5 files there. These thousands of locus dirs are created inside the temporary dir. It would be too memory heavy to keep all this data in RAM, but maybe we could do some streaming process such that we don't need to write all these files, and later combine them. This might work, but it is a good amount of work just for not creating temp dirs/files, which is not a big issue. 2. `MAFFT` necessarily needs a temp dir (and possibly future aligners). 3. I wanted to control where the temp dir is created i.e. in the output folder, because I see many pipelines/runs failing because the space in the temp/scratch folder is full, especially in clusters. Sure, the standard is to create temp files in the standard temp dir (by default `/tmp`), and users can always override this default with `TMPDIR`, but I think creating temp dirs/files in the output dir will solve some headaches and issues in the future. This is my POV though, would be nice to know if you disagree with any of these points and propose an alternative solution.

Reply:
I'm not arguing that we don't need temporary directories - we absolutely do. I was just questioning why you are manually creating unique names etc. when `tempfile` handles this exact functionality - albeit thorough tested and supported. You can also tell `tempfile` *where* to create the temporary directories and files - it need not be `TMPDIR`.

Anyway, I'm not suggesting you make drastic changes. I was just flagging this as something that might be worth migrating to in the future and any new temporary file/dir functionality should maybe be handled by `tempfile`. Also, it would mean a lot less code to test.

Reply:
> I'm not arguing that we don't need temporary directories - we absolutely do. I was just questioning why you are manually creating unique names etc. when tempfile handles this exact functionality - albeit thorough tested and supported. You can also tell tempfile where to create the temporary directories and files - it need not be TMPDIR.

Yeah, just found this out when you linked the docs to tempfile, and implemented it here: https://github.com/iqbal-lab-org/make_prg/pull/34/commits/ee0a383713c81630fc7505e6d901c07da83d76a6

> Anyway, I'm not suggesting you make drastic changes. I was just flagging this as something that might be worth migrating to in the future and any new temporary file/dir functionality should maybe be handled by tempfile. Also, it would mean a lot less code to test.

Yep, has already implemented your suggestion here: https://github.com/iqbal-lab-org/make_prg/pull/34/commits/ee0a383713c81630fc7505e6d901c07da83d76a6